server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -58,42 +58,54 @@ public final class IngestDocument {
     private boolean doNoSelfReferencesCheck = false;
 
     public IngestDocument(String index, String id, long version, String routing, VersionType versionType, Map<String, Object> source) {
-        // source + at max 5 extra fields
-        this.sourceAndMetadata = Maps.newMapWithExpectedSize(source.size() + 5);
-        this.sourceAndMetadata.putAll(source);
-        this.sourceAndMetadata.put(Metadata.INDEX.getFieldName(), index);
-        this.sourceAndMetadata.put(Metadata.ID.getFieldName(), id);
-        this.sourceAndMetadata.put(Metadata.VERSION.getFieldName(), version);
-        if (routing != null) {
-            this.sourceAndMetadata.put(Metadata.ROUTING.getFieldName(), routing);
-        }
-        if (versionType != null) {
-            sourceAndMetadata.put(Metadata.VERSION_TYPE.getFieldName(), VersionType.toString(versionType));
-        }
+        this.sourceAndMetadata = new IngestSourceAndMetadata(
+            index,
+            id,
+            version,
+            routing,
+            versionType,
+            ZonedDateTime.now(ZoneOffset.UTC),
+            source
+        );
         this.ingestMetadata = new HashMap<>();
-        this.ingestMetadata.put(TIMESTAMP, ZonedDateTime.now(ZoneOffset.UTC));
+        this.ingestMetadata.put(TIMESTAMP, sourceAndMetadata.getTimestamp());
     }
 
     /**
      * Copy constructor that creates a new {@link IngestDocument} which has exactly the same properties as the one provided as argument
      */
     public IngestDocument(IngestDocument other) {
-        this(deepCopyMap(other.sourceAndMetadata), deepCopyMap(other.ingestMetadata));
+        this(
+            new IngestSourceAndMetadata(
+                deepCopyMap(other.sourceAndMetadata.getSource()),
+                deepCopyMap(other.sourceAndMetadata.getMetadata()),
+                other.getIngestSourceAndMetadata().timestamp,
+                other.getIngestSourceAndMetadata().validators
+            ),
+            deepCopyMap(other.ingestMetadata)
+        );
     }
 
     /**
-     * Constructor to create an IngestDocument from its constituent maps
+     * Constructor to create an IngestDocument from its constituent maps.  The maps are shallow copied.
      */
-    IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, Object> ingestMetadata) {
-        this.sourceAndMetadata = sourceAndMetadata;
-        this.ingestMetadata = ingestMetadata;
+    public IngestDocument(Map<String, Object> sourceAndMetadata, Map<String, Object> ingestMetadata) {
+        Tuple<Map<String, Object>, Map<String, Object>> sm = IngestSourceAndMetadata.splitSourceAndMetadata(sourceAndMetadata);
+        this.sourceAndMetadata = new IngestSourceAndMetadata(
+            sm.v1(),
+            sm.v2(),
+            IngestSourceAndMetadata.getTimestamp(ingestMetadata),
+            IngestSourceAndMetadata.VALIDATORS
+        );
+        this.ingestMetadata = ingestMetadata != null ? ingestMetadata : new HashMap<>();
```

Comment:
Doesn't this need to be the metadata from sourceAndMetadata? If the assumption is it should already exist in the passed in sourceAndMetadata, then shouldn't ingestMetadata always be non null? The naming here is a bit confusing because "sourceAndMetadata" implies it contains the metadata, but then ingestMetadata is *also* passed in. When this was package private it's scoped more narrowly, but making this ctor public opens this up to trappy usage.

Reply:
Removed `HashMap` instantiation on `null`.